examples/file-sharing.rs

```diff
@@ -0,0 +1,639 @@
+use libp2p::multiaddr::Protocol;
+use std::error::Error;
+use std::path::PathBuf;
+use structopt::StructOpt;
+
+#[async_std::main]
+async fn main() -> Result<(), Box<dyn Error>> {
+    env_logger::init();
+
+    let opt = Opt::from_args();
+
+    // Creates the network components, namely:
+    //
+    // - The network client to interact with the network layer from anywhere
+    //   within your application.
+    //
+    // - The network event stream, e.g. for incoming requests.
+    //
+    // - The network task driving the network itself.
+    let (mut network_client, mut network_events, network_task) =
+        network::new(opt.listen_address, opt.secret_key_seed)?;
+
+    // Spawn the network task for it to run in the background.
+    spawn(network_task);
+
+    // In case the user provided an address of a peer on the CLI, dial it.
+    if let Some(addr) = opt.peer {
+        let peer_id = match addr.iter().last() {
+            Some(Protocol::P2p(hash)) => PeerId::from_multihash(hash).expect("Valid hash."),
+            _ => return Err("Expect peer multiaddr to contain peer ID.".into()),
+        };
+        network_client.dial(peer_id, addr).await;
+    }
+
+    match opt.command {
+        // Providing a file.
+        Command::Provide { path, name } => {
+            // Advertise oneself as a provider of the file on the DHT.
+            network_client.start_providing(name.clone()).await;
+
+            loop {
+                match network_events.next().await {
+                    // Reply with the content of the file on incoming requests.
+                    Some(network::Event::InboundRequest { request, channel }) => {
+                        if request == name {
+                            let file_content = std::fs::read_to_string(&path)?;
+                            network_client.respond_file(file_content, channel).await;
+                        }
+                    }
+                    _ => todo!(),
+                }
+            }
+        }
+        // Locating and getting a file.
+        Command::Get { name } => {
+            // Locate all nodes providing the file.
+            let providers = network_client.get_providers(name.clone()).await;
+            if providers.is_empty() {
+                return Err(format!("Could not find provider for file {}.", name).into());
+            }
+
+            // Request the content of the file from each node.
+            let requests = providers.into_iter().map(|p| {
+                let mut network_client = network_client.clone();
+                let name = name.clone();
+                async move { network_client.request_file(p, name).await.ok_or(()) }.boxed()
+            });
+
+            // Await the requests, ignore the remaining once a single one succeeds.
+            let file = futures::future::select_ok(requests)
+                .await
+                .map_err(|_| Into::<Box<dyn Error>>::into("None of the providers returned file."))?
```

Comment:
With `anyhow`, this would be a simple `.context`.

Reply:
It is just the trait off between better error handling and yet another dependency increasing compile times / CI times. Given that the main purpose of this example is not showcasing great error handling, I went without `anyhow`. I cleaned up the `map_err` a bit now though.